File: Baekjoon/1182.py
import sys
input = sys.stdin.readline

n , s = map(int, input().split())
arr = list(map(int, input().split()))
cnt = 0

def dfs(idx, sum_result):
  global cnt
  # 인덱스가 n을 넘어가는 경우는 바로 리턴
  if idx >= n:
    return

  sum_result += arr[idx]

  if sum_result == s: # 현재 더한값이 s랑 같은 경우
    cnt += 1

  dfs(idx + 1, sum_result - arr[idx]) # 자기 자신을 더하지 않은 경우
  dfs(idx + 1, sum_result) # 자기 자신을 더한 경우

dfs(0,0)

print(cnt)

File: Baekjoon/1182.java
import java.util.*;
import java.io.*;

public class Main {
    static int n, s;
    static int cnt;
    static int sum_result = 0;
    static int idx = 0;
    static int[] arr;
    
    public static void main(String args[]) throws IOException{
      BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
      StringTokenizer st = new StringTokenizer(br.readLine());
      n = Integer.parseInt(st.nextToken());
      s = Integer.parseInt(st.nextToken());
      
      arr = new int[n];
      st = new StringTokenizer(br.readLine());
      
      for(int i=0; i<n; i++){
          arr[i] = Integer.parseInt(st.nextToken());
      }
      
      dfs(idx, sum_result);
      
      System.out.println(cnt);
    }
    
    static void dfs(int idx, int sum){
        if (idx >= n)
            return;
            
        sum += arr[idx];
        
        if (sum == s)
            cnt++;
            
        dfs(idx + 1, sum - arr[idx]);
        dfs(idx + 1, sum);
    }
}
